internal/internal_workflow_client.go

```diff
@@ -1170,3 +950,273 @@ func serializeSearchAttributes(input map[string]interface{}) (*commonpb.SearchAt
+	if err != nil {
+		return nil, err
+	}
+
+	// run propagators to extract information about tracing and other stuff, store in headers field
+	startRequest := &workflowservice.StartWorkflowExecutionRequest{
+		Namespace:                w.client.namespace,
+		RequestId:                uuid.New(),
+		WorkflowId:               workflowID,
+		WorkflowType:             &commonpb.WorkflowType{Name: in.WorkflowType},
+		TaskQueue:                &taskqueuepb.TaskQueue{Name: in.Options.TaskQueue, Kind: enumspb.TASK_QUEUE_KIND_NORMAL},
+		Input:                    input,
+		WorkflowExecutionTimeout: &executionTimeout,
+		WorkflowRunTimeout:       &runTimeout,
+		WorkflowTaskTimeout:      &workflowTaskTimeout,
+		Identity:                 w.client.identity,
+		WorkflowIdReusePolicy:    in.Options.WorkflowIDReusePolicy,
+		RetryPolicy:              convertToPBRetryPolicy(in.Options.RetryPolicy),
+		CronSchedule:             in.Options.CronSchedule,
+		Memo:                     memo,
+		SearchAttributes:         searchAttr,
+		Header:                   header,
+	}
+
+	var response *workflowservice.StartWorkflowExecutionResponse
+
+	grpcCtx, cancel := newGRPCContext(ctx, grpcMetricsScope(
+		metrics.GetMetricsScopeForRPC(w.client.metricsScope, in.WorkflowType, metrics.NoneTagValue, in.Options.TaskQueue)),
+		defaultGrpcRetryParameters(ctx))
+	defer cancel()
+
+	response, err = w.client.workflowService.StartWorkflowExecution(grpcCtx, startRequest)
+
+	// Allow already-started error
+	var runID string
+	if e, ok := err.(*serviceerror.WorkflowExecutionAlreadyStarted); ok && !in.Options.WorkflowExecutionErrorWhenAlreadyStarted {
+		runID = e.RunId
+	} else if err != nil {
+		return nil, err
+	} else {
+		runID = response.RunId
+	}
+
+	iterFn := func(fnCtx context.Context, fnRunID string) HistoryEventIterator {
+		rpcScope := metrics.GetMetricsScopeForRPC(w.client.metricsScope, in.WorkflowType,
+			metrics.NoneTagValue, in.Options.TaskQueue)
+		return w.client.getWorkflowHistory(fnCtx, workflowID, fnRunID, true,
+			enumspb.HISTORY_EVENT_FILTER_TYPE_CLOSE_EVENT, rpcScope)
+	}
+
+	curRunIDCell := util.PopulatedOnceCell(runID)
+	return &workflowRunImpl{
+		workflowType:  in.WorkflowType,
+		workflowID:    workflowID,
+		firstRunID:    runID,
+		currentRunID:  &curRunIDCell,
+		iterFn:        iterFn,
+		dataConverter: w.client.dataConverter,
+		registry:      w.client.registry,
+	}, nil
+}
+
+func (w *workflowClientInterceptor) SignalWorkflow(ctx context.Context, in *ClientSignalWorkflowInput) error {
+	dataConverter := WithContext(ctx, w.client.dataConverter)
+	input, err := encodeArg(dataConverter, in.Arg)
+	if err != nil {
+		return err
+	}
+
+	request := &workflowservice.SignalWorkflowExecutionRequest{
+		Namespace: w.client.namespace,
+		WorkflowExecution: &commonpb.WorkflowExecution{
+			WorkflowId: in.WorkflowID,
+			RunId:      in.RunID,
+		},
+		SignalName: in.SignalName,
+		Input:      input,
+		Identity:   w.client.identity,
```

Comment:
we should set RequestId here, I realize that we didn't set it in old code either, but that is not right.

Reply:
I don't think this PR is the place for such fixes. There are at least 5 places where we set the `Identity` proto fields of requests as the `client.identity`. Can you open an issue detailing the problem and solution?

Reply:
https://github.com/temporalio/sdk-go/issues/627